source/code/go/src/plugins/telemetry.go

```diff
@@ -0,0 +1,155 @@
+		return -1, errors.New("Missing Environment Variable")
+	}
+
+	decIkey, err := base64.StdEncoding.DecodeString(encodedIkey)
+	if err != nil {
+		Log("Decoding Error %s", err.Error())
+		return -1, err
+	}
+
+	TelemetryClient = appinsights.NewTelemetryClient(string(decIkey))
+
+	CommonProperties = make(map[string]string)
+	CommonProperties["Computer"] = Computer
+	CommonProperties["WorkspaceID"] = WorkspaceID
+	CommonProperties["ControllerType"] = controllerTypeDaemonSet
+	CommonProperties["AgentVersion"] = agentVersion
+
+	aksResourceID := os.Getenv(envAKSResourceID)
+	// if the aks resource id is not defined, it is most likely an ACS Cluster
+	if aksResourceID == "" {
+		CommonProperties["ACSResourceName"] = os.Getenv(envACSResourceName)
+		CommonProperties["ClusterType"] = clusterTypeACS
+
+		CommonProperties["SubscriptionID"] = ""
+		CommonProperties["ResourceGroupName"] = ""
+		CommonProperties["ClusterName"] = ""
+		CommonProperties["Region"] = ""
+
+	} else {
+		CommonProperties["ACSResourceName"] = ""
+		splitStrings := strings.Split(aksResourceID, "/")
+		CommonProperties["SubscriptionID"] = splitStrings[2]
+		CommonProperties["ResourceGroupName"] = splitStrings[4]
+		CommonProperties["ClusterName"] = splitStrings[8]
+		CommonProperties["ClusterType"] = clusterTypeAKS
+
+		region := os.Getenv("AKS_REGION")
+		if region != "" {
+			CommonProperties["Region"] = region
+		}
+	}
+
+	TelemetryClient.Context().CommonProperties = CommonProperties
+	return 0, nil
+}
+
+// SendContainerLogFlushRateMetric is a go-routine that flushes the data periodically (every 5 mins to App Insights)
+func SendContainerLogFlushRateMetric(telemetryPushIntervalProperty string, agentVersion string) {
+
+	ret, err := initialize(telemetryPushIntervalProperty, agentVersion)
+	if ret != 0 || err != nil {
+		Log("Error During Telemetry Initialization :%s", err.Error())
+		runtime.Goexit()
+	}
+
+	SendEvent(EventNameContainerLogInit, make(map[string]string))
+
+	for ; true; <-ContainerLogTelemetryTicker.C {
+		flushRate := FlushedRecordsCount / FlushedRecordsTimeTaken * 1000
+		metric := appinsights.NewMetricTelemetry(metricNameAvgFlushRate, flushRate)
+		Log("Flushed Records : %f Time Taken : %f flush Rate : %f", FlushedRecordsCount, FlushedRecordsTimeTaken, flushRate)
+		TelemetryClient.Track(metric)
+		FlushedRecordsCount = 0.0
+		FlushedRecordsTimeTaken = 0.0
+	}
+}
+
+// TelemetryShutdown stops the ticker that sends data to App Insights periodically
+func TelemetryShutdown() {
+	Log("Shutting down ContainerLog Telemetry\n")
+	ContainerLogTelemetryTicker.Stop()
+}
+
+// SendEvent sends an event to App Insights
+func SendEvent(eventName string, dimensions map[string]string) {
+	// this is because the TelemetryClient is initialized in a different goroutine. A simple wait loop here is just waiting for it to be initialized. This will happen only for the init event. Any subsequent Event should work just fine
+	for TelemetryClient == nil {
```

Comment:
this is not ideal (waiting on an infintite for loop). Can we init & do all telemetry in the same go routine may be ?